compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Optics.hs

```diff
@@ -133,7 +127,10 @@ instance {-# OVERLAPPING #-} MonoTraversable ModuleRef FilePath where monoTraver
 instance MonoTraversable ModuleRef Kind where monoTraverse _ = pure
 instance MonoTraversable ModuleRef BuiltinType where monoTraverse _ = pure
 instance MonoTraversable ModuleRef BuiltinExpr where monoTraverse _ = pure
-instance MonoTraversable ModuleRef SourceLoc where monoTraverse _ = pure
+
+instance MonoTraversable ModuleRef SourceLoc where
+  monoTraverse f sl@SourceLoc{slocModuleRef} =
+    (\mr -> sl {slocModuleRef = mr}) <$> traverse f slocModuleRef
```

Comment:
@dajmaki Was there any reason for the prior instance? See https://github.com/digital-asset/daml/pull/1996#discussion_r300250409

Reply:
We use `monoTraverse` to collect all `ModuleRefs` in a module or package in order to figure out its dependencies. Inlining can cause location information to reference the original source file although there's not a proper dependency. Given that we're not using GHC's inliner, I'm not sure we ever hit this case. Even if we do, it shouldn't be too often and not interning here shouldn't do any harm. Thus, I'd prefer if we don't change this and give a better explanation why we don't do it.

Reply:
It's a little more concerning in the context of LF language 2.x, where we should eliminate the uninterned case of `PackageRef` entirely.

Reply:
Not-doing and explanation in 6312b987bf944c7170a9bea396fb5a14c35daf65

Reply:
cc @cocreature who wondered why they were skipped

Reply:
Not entirely convinced by the explanation tbh. I wouldn’t expect interning to randomly (in the context of interning) skip things just because we want to reuse a traversal. But since this PR doesn’t change this, I’m fine with it for now.

Reply:
Actually, looking at this again. It seems like this is exactly what this PR changes. So I’m not sure the changes here are actually a good idea.
I would change `MonoTraversable` to hit all `PackageRef`s and then have a separate traversal for collecting dependencies that skips `SourceLoc`.

Reply:
I agree that this situation is not optimal. I'm tempted to merge this now nevertheless and clean up the mess in a future PR.

Reply:
To be more specific, we skip the `ModuleRef` in `SourceLoc` when collecting dependencies for correctness reasons whereas we'd like to intern it for for performance reasons. Now, the former should win but in the long run we need to do both of them properly.

Reply:
Without this PR, we skipped `ModuleRef` when collecting dependencies and we didn’t skip it when interning. So this PR does improve neither correctness nor performance, in fact it’s a regression in performance and I would argue that skipping things during interning does at least break the very reasonable assumption that everything is interned afterwards.

Reply:
> So this PR does improve neither correctness nor performance, in fact it’s a regression in performance

@cocreature You're right, but for now, the small (possibly nil as @hurryabit said) likelihood of not interning a SourceLoc PackageId because it's not referenced in a non-SourceLoc context anywhere in the package isn't worth having code for two traversals.

Note that not traversing a package ref doesn't necessarily mean it won't be interned. For example, I compared the runs before and after this patch on a very large LF package, and the size was exactly the same, so any external SourceLocs therein still had their package IDs interned.

> and I would argue that skipping things during interning does at least break the very reasonable assumption that everything is interned afterwards.

We deliberately didn't promise to intern everything, because the LF governance rules and [testing tools](https://github.com/digital-asset/daml/blob/0c154b7948ab28f21730962cb82e08c294d99c42/daml-lf/encoder/src/main/scala/com/digitalasset/daml/lf/archive/testing/EncodeV1.scala#L78-L82) make it much simpler to reserve that promise for LF v2.

> I would change `MonoTraversable` to hit all `PackageRef`s and then have a separate traversal for collecting dependencies that skips `SourceLoc`.

I gave the `MonoTraversable` priority for its seniority. However, for 2 traversals, I would prefer to delete `MonoTraversable`, resurrecting and parameterizing the `Generic`-derived traversal (we can't change `monoTraverse`'s signature) by whether you want to skip `SourceLoc`.